Mathlib/RingTheory/MvPolynomial/Homogeneous.lean

```diff
@@ -230,6 +232,119 @@ theorem rename_isHomogeneous_iff {f : σ → τ} (hf : f.Injective) :
+
+section domain
+
+-- In this section we shadow the semiring `R` with a domain `R`.
+variable {R σ : Type*} [CommRing R] [IsDomain R] {F : MvPolynomial σ R} {n : ℕ}
+
+open Cardinal Polynomial
+private
+lemma exists_eval_ne_zero_of_totalDegree_le_card_aux {N : ℕ} {F : MvPolynomial (Fin N) R} {n : ℕ}
+    (hF : F.IsHomogeneous n) (hF₀ : F ≠ 0) (hnR : n ≤ #R) :
+    ∃ r, eval r F ≠ 0 := by
+  induction N generalizing n with
+  | zero =>
+    use 0
+    contrapose! hF₀
+    ext d
+    simpa only [Subsingleton.elim d 0, eval_zero, coeff_zero] using hF₀
+  | succ N IH =>
+    have hdeg : natDegree (finSuccEquiv R N F) < n + 1 := by
+      linarith [natDegree_finSuccEquiv F, degreeOf_le_totalDegree F 0, hF.totalDegree hF₀]
+    obtain ⟨i, hi⟩ : ∃ i : ℕ, (finSuccEquiv R N F).coeff i ≠ 0 := by
+      contrapose! hF₀
+      exact (finSuccEquiv _ _).injective <| Polynomial.ext <| by simpa using hF₀
+    have hin : i ≤ n := by
+      contrapose! hi
+      exact coeff_eq_zero_of_natDegree_lt <| (Nat.le_of_lt_succ hdeg).trans_lt hi
+    obtain hFn | hFn := ne_or_eq ((finSuccEquiv R N F).coeff n) 0
+    · exact hF.exists_eval_ne_zero_of_totalDegree_le_card_aux₀ hdeg hFn
+    have hin : i < n := hin.lt_or_eq.elim id <| by aesop
+    obtain ⟨j, hj⟩ : ∃ j, i + (j + 1) = n := (Nat.exists_eq_add_of_lt hin).imp <| by intros; omega
+    obtain ⟨r, hr⟩ : ∃ r, (eval r) (Polynomial.coeff ((finSuccEquiv R N) F) i) ≠ 0 :=
+      IH (hF.finSuccEquiv_coeff_isHomogeneous _ _ hj) hi (.trans (by norm_cast; omega) hnR)
+    set φ : R[X] := Polynomial.map (eval r) (finSuccEquiv _ _ F) with hφ
+    have hφ₀ : φ ≠ 0 := fun hφ₀ ↦ hr <| by
+      rw [← coeff_eval_eq_eval_coeff, ← hφ, hφ₀, Polynomial.coeff_zero]
+    have hφR : φ.natDegree < #R := by
+      refine lt_of_lt_of_le ?_ hnR
+      norm_cast
+      refine lt_of_le_of_lt (natDegree_map_le _ _) ?_
+      suffices (finSuccEquiv _ _ F).natDegree ≠ n by omega
+      rintro rfl
+      refine leadingCoeff_ne_zero.mpr ?_ hFn
+      simpa using (finSuccEquiv R N).injective.ne hF₀
+    obtain ⟨r₀, hr₀⟩ : ∃ r₀, Polynomial.eval r₀ φ ≠ 0 :=
+      φ.exists_eval_ne_zero_of_natDegree_lt_card hφ₀ hφR
+    use Fin.cons r₀ r
+    rwa [eval_eq_eval_mv_eval']
+
+open Cardinal in
+lemma exists_eval_ne_zero_of_totalDegree_le_card
+    (hF : F.IsHomogeneous n) (hF₀ : F ≠ 0) (h : n ≤ #R) :
+    ∃ r : σ → R, eval r F ≠ 0 := by
+  -- reduce to the case where σ is finite
+  obtain ⟨k, f, hf, F, rfl⟩ := exists_fin_rename F
+  have hF₀ : F ≠ 0 := by rintro rfl; simp at hF₀
+  have hF : F.IsHomogeneous n := by rwa [rename_isHomogeneous_iff hf] at hF
+  obtain ⟨r, hr⟩ := exists_eval_ne_zero_of_totalDegree_le_card_aux hF hF₀ h
+  obtain ⟨r, rfl⟩ := (Function.factorsThrough_iff _).mp <| (hf.factorsThrough r)
+  use r
+  rwa [eval_rename]
+
+open Cardinal in
+/-- See `MvPolynomial.IsHomogeneous.funext` for a version that assumes `Infinite R`. -/
+lemma funext_of_le_card (hF : F.IsHomogeneous n) (hF₀ : ∀ r : σ → R, eval r F = 0) (h : n ≤ #R) :
```

Comment:
I think we should probably reserve the name `funext` for a lemma stating `F = G`. How about:
```suggestion
/-- See `MvPolynomial.IsHomogeneous.eq_zero_of_forall_eval_eq_zero` for a version that assumes
`Infinite R`. -/
lemma eq_zero_of_forall_eval_eq_zero_of_le_card
    (hF : F.IsHomogeneous n) (hF₀ : ∀ r : σ → R, eval r F = 0) (h : n ≤ #R) :
```
For that matter, it would be nice if we had:
```lean
@[ext]
lemma funext {G : MvPolynomial σ R} [Infinite R]
    (hF : F.IsHomogeneous n) (hG : G.IsHomogeneous n)
    (h : ∀ r, eval r F = eval r G) : F = G := by
  sorry
```
but this is not necessary (and I'm not 100% about that `@[ext]` attribute --- I always test them before I commit).

Reply:
We already have the `ext` for "coefficient-wise equality". And I think that one should be the default.

But yes, I will add this lemmas.

Reply:
I see, with this in mind and the new additions, these lemmas look great now. I did wonder if we need the `F = 0` variants since we have the `F = G` versions but I can imagine these being quite useful for rewriting.